File: app/jobs/recalculate_contact_metrics_job.rb
# frozen_string_literal: true

# RecalculateContactMetricsJob
#
# Background job to recalculate fingerprints and quality scores for contacts
# after bulk imports that skipped callbacks.
#
# This job processes contacts in batches to prevent memory issues and
# provides progress tracking for long-running operations.
#
# Usage:
#   # After bulk import with callbacks skipped:
#   RecalculateContactMetricsJob.perform_later(contact_ids)
#
#   # Process all contacts (expensive - use only for maintenance):
#   RecalculateContactMetricsJob.perform_later(Contact.pluck(:id))
#
class RecalculateContactMetricsJob < ApplicationJob
  queue_as :low_priority

  # Retry on database errors (transient failures)
  retry_on ActiveRecord::Deadlocked, wait: :exponentially_longer, attempts: 3
  retry_on ActiveRecord::StatementInvalid, wait: :exponentially_longer, attempts: 3

  # Don't retry if contact was deleted
  discard_on ActiveRecord::RecordNotFound

  # Process in batches to prevent memory issues
  BATCH_SIZE = 100

  def perform(contact_ids, batch_index: 0)
    return if contact_ids.blank?

    Rails.logger.info("RecalculateContactMetricsJob: Processing batch #{batch_index + 1}, " \
                      "#{contact_ids.size} contacts total")

    # Process this batch
    current_batch = contact_ids.first(BATCH_SIZE)
    remaining = contact_ids.drop(BATCH_SIZE)

    process_batch(current_batch)

    # Enqueue next batch if there are more contacts
    if remaining.any?
      self.class.perform_later(remaining, batch_index: batch_index + 1)
      Rails.logger.info("RecalculateContactMetricsJob: Enqueued batch #{batch_index + 2} " \
                        "with #{remaining.size} remaining contacts")
    else
      Rails.logger.info("RecalculateContactMetricsJob: All batches completed " \
                        "(#{batch_index + 1} total batches)")
    end
  end

  private

  def process_batch(contact_ids)
    start_time = Time.current
    updated_count = 0

    Contact.where(id: contact_ids).find_each do |contact|
      # Update fingerprints for duplicate detection
      contact.update_fingerprints!

      # Recalculate quality score
      contact.calculate_quality_score!

      updated_count += 1
    rescue StandardError => e
      # Log error but continue processing other contacts
      Rails.logger.error(
        "RecalculateContactMetricsJob: Failed to update contact #{contact.id}: #{e.message}"
      )
    end

    duration = Time.current - start_time
    avg_time = duration / updated_count if updated_count.positive?

    Rails.logger.info(
      "RecalculateContactMetricsJob: Batch complete - " \
      "#{updated_count}/#{contact_ids.size} contacts updated in #{duration.round(2)}s " \
      "(avg: #{avg_time&.round(3)}s per contact)"
    )
  end
end

File: laravel-app/app/Jobs/RecalculateContactMetricsJob.php
<?php

namespace App\Jobs;

use App\Models\Contact;
use Illuminate\Bus\Queueable;
use Illuminate\Contracts\Queue\ShouldQueue;
use Illuminate\Foundation\Bus\Dispatchable;
use Illuminate\Queue\InteractsWithQueue;
use Illuminate\Queue\SerializesModels;
use Illuminate\Support\Facades\Log;
use Illuminate\Database\QueryException;
use Exception;

/**
 * RecalculateContactMetricsJob
 *
 * Background job to recalculate fingerprints and quality scores for contacts
 * after bulk imports that skipped callbacks.
 *
 * This job processes contacts in batches to prevent memory issues and
 * provides progress tracking for long-running operations.
 *
 * Usage:
 *   // After bulk import with callbacks skipped:
 *   RecalculateContactMetricsJob::dispatch($contactIds);
 *
 *   // Process all contacts (expensive - use only for maintenance):
 *   RecalculateContactMetricsJob::dispatch(Contact::pluck('id')->toArray());
 */
class RecalculateContactMetricsJob implements ShouldQueue
{
    use Dispatchable, InteractsWithQueue, Queueable, SerializesModels;

    public $contactIds;
    public $batchIndex;

    /**
     * The number of times the job may be attempted.
     *
     * @var int
     */
    public $tries = 3;

    /**
     * The number of seconds to wait before retrying.
     *
     * @var int
     */
    public $backoff = [10, 30, 90];

    /**
     * The maximum number of seconds the job can run before timing out.
     *
     * @var int
     */
    public $timeout = 600;

    /**
     * Process in batches to prevent memory issues
     */
    const BATCH_SIZE = 100;

    /**
     * Create a new job instance.
     *
     * @param array $contactIds
     * @param int $batchIndex
     */
    public function __construct(array $contactIds, int $batchIndex = 0)
    {
        $this->contactIds = $contactIds;
        $this->batchIndex = $batchIndex;
        $this->onQueue('low_priority');
    }

    /**
     * Execute the job.
     *
     * @return void
     */
    public function handle()
    {
        if (empty($this->contactIds)) {
            return;
        }

        Log::info("RecalculateContactMetricsJob: Processing batch " . ($this->batchIndex + 1) . ", " .
            count($this->contactIds) . " contacts total");

        // Process this batch
        $currentBatch = array_slice($this->contactIds, 0, self::BATCH_SIZE);
        $remaining = array_slice($this->contactIds, self::BATCH_SIZE);

        $this->processBatch($currentBatch);

        // Enqueue next batch if there are more contacts
        if (count($remaining) > 0) {
            self::dispatch($remaining, $this->batchIndex + 1);
            Log::info("RecalculateContactMetricsJob: Enqueued batch " . ($this->batchIndex + 2) .
                " with " . count($remaining) . " remaining contacts");
        } else {
            Log::info("RecalculateContactMetricsJob: All batches completed " .
                "(" . ($this->batchIndex + 1) . " total batches)");
            // Ensure dashboard stats reflect recalculated metrics
            DashboardBroadcastJob::dispatch();
        }
    }

    /**
     * Process a batch of contacts
     *
     * @param array $contactIds
     * @return void
     */
    protected function processBatch(array $contactIds)
    {
        $startTime = now();
        $updatedCount = 0;

        foreach (Contact::whereIn('id', $contactIds)->cursor() as $contact) {
            try {
                // Update fingerprints for duplicate detection
                $contact->updateFingerprints();

                // Recalculate quality score
                $contact->calculateQualityScore();

                $updatedCount++;
            } catch (Exception $e) {
                // Log error but continue processing other contacts
                Log::error("RecalculateContactMetricsJob: Failed to update contact {$contact->id}: {$e->getMessage()}");
            }
        }

        $duration = now()->diffInSeconds($startTime);
        $avgTime = $updatedCount > 0 ? $duration / $updatedCount : 0;

        Log::info("RecalculateContactMetricsJob: Batch complete - " .
            "{$updatedCount}/" . count($contactIds) . " contacts updated in {$duration}s " .
            "(avg: " . round($avgTime, 3) . "s per contact)");
    }

    /**
     * Handle a job failure.
     *
     * @param  \Throwable  $exception
     * @return void
     */
    public function failed(\Throwable $exception)
    {
        Log::error("RecalculateContactMetricsJob failed for batch {$this->batchIndex}: {$exception->getMessage()}");
    }
}
